packages/runtime/container-runtime/src/opLifecycle/opDecompressor.ts

```diff
@@ -122,3 +119,34 @@ const newMessage = (
 	compression: undefined,
 	metadata: { ...originalMessage.metadata },
 });
+
+const isCompressed = (message: ISequencedDocumentMessage): boolean => {
+	if (message.compression === CompressionAlgorithms.lz4) {
+		return true;
+	}
+
+	// This condition holds true for compressed messages, regardless of metadata.
+	// Back-compat self healing mechanism for ADO:3538, as loaders from
+	// version client_v2.0.0-internal.1.2.0 to client_v2.0.0-internal.2.2.0 do not
+	// support adding the proper compression metadata to compressed messages submitted
+	// by the runtime. Should be removed after the loader reaches sufficient saturation
+	// for a version greater or equal than client_v2.0.0-internal.2.2.0.
+	if (typeof message.contents === "object" && message.contents?.packedContents !== undefined) {
+		return (
+			Object.keys(message.contents).length === 1 &&
+			typeof message.contents?.packedContents === "string" &&
+			message.contents.packedContents.length > 0 &&
+			isBase64(message.contents.packedContents)
+		);
+	}
+
+	return false;
+};
+
+const isBase64 = (content: string): boolean => {
+	try {
```

Comment:
I'd rather inline this fuction and wrap with try/catch much bigger piece of code. JS spec is really weird, we could easily miss something about it (for example, maybe Object.keys() will through with some type of object.
For example if message.contents === null, it will pass first check, but explode on Object.keys()